File: android/src/main/java/com/mostlygoodmetrics/reactnative/MostlyGoodMetricsModule.kt
package com.mostlygoodmetrics.reactnative

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.mostlygoodmetrics.sdk.MGMConfiguration
import com.mostlygoodmetrics.sdk.MostlyGoodMetrics

class MostlyGoodMetricsModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = "MostlyGoodMetricsModule"

    @ReactMethod
    fun configure(apiKey: String, options: ReadableMap) {
        val builder = MGMConfiguration.Builder(apiKey)

        if (options.hasKey("baseURL")) {
            builder.baseUrl(options.getString("baseURL")!!)
        }
        if (options.hasKey("environment")) {
            builder.environment(options.getString("environment")!!)
        }
        if (options.hasKey("bundleId")) {
            builder.packageName(options.getString("bundleId")!!)
        }
        if (options.hasKey("maxBatchSize")) {
            builder.maxBatchSize(options.getInt("maxBatchSize"))
        }
        if (options.hasKey("flushInterval")) {
            builder.flushIntervalSeconds(options.getInt("flushInterval").toLong())
        }
        if (options.hasKey("maxStoredEvents")) {
            builder.maxStoredEvents(options.getInt("maxStoredEvents"))
        }
        if (options.hasKey("enableDebugLogging")) {
            builder.enableDebugLogging(options.getBoolean("enableDebugLogging"))
        }
        if (options.hasKey("trackAppLifecycleEvents")) {
            builder.trackAppLifecycleEvents(options.getBoolean("trackAppLifecycleEvents"))
        }

        MostlyGoodMetrics.configure(reactApplicationContext, builder.build())
    }

    @ReactMethod
    fun track(name: String, properties: ReadableMap?) {
        val props = properties?.toHashMap()?.mapValues { it.value }
        MostlyGoodMetrics.track(name, props)
    }

    @ReactMethod
    fun identify(userId: String) {
        MostlyGoodMetrics.identify(userId)
    }

    @ReactMethod
    fun resetIdentity() {
        MostlyGoodMetrics.resetIdentity()
    }

    @ReactMethod
    fun flush() {
        MostlyGoodMetrics.flush()
    }

    @ReactMethod
    fun startNewSession() {
        MostlyGoodMetrics.shared?.startNewSession()
    }

    @ReactMethod
    fun clearPendingEvents() {
        MostlyGoodMetrics.shared?.clearPendingEvents()
    }

    @ReactMethod
    fun getPendingEventCount(promise: Promise) {
        val count = MostlyGoodMetrics.shared?.pendingEventCount ?: 0
        promise.resolve(count)
    }
}

File: ios/MostlyGoodMetricsModule.swift
import Foundation
import MostlyGoodMetrics

@objc(MostlyGoodMetricsModule)
class MostlyGoodMetricsModule: NSObject {

    @objc
    static func requiresMainQueueSetup() -> Bool {
        return false
    }

    @objc
    func configure(_ apiKey: String, options: NSDictionary) {
        var config = MGMConfiguration(apiKey: apiKey)

        if let baseURL = options["baseURL"] as? String, let url = URL(string: baseURL) {
            config = MGMConfiguration(
                apiKey: apiKey,
                baseURL: url,
                environment: options["environment"] as? String ?? "production",
                bundleId: options["bundleId"] as? String,
                maxBatchSize: options["maxBatchSize"] as? Int ?? 100,
                flushInterval: options["flushInterval"] as? TimeInterval ?? 30,
                maxStoredEvents: options["maxStoredEvents"] as? Int ?? 10000,
                enableDebugLogging: options["enableDebugLogging"] as? Bool ?? false,
                trackAppLifecycleEvents: options["trackAppLifecycleEvents"] as? Bool ?? true
            )
        } else {
            config = MGMConfiguration(
                apiKey: apiKey,
                environment: options["environment"] as? String ?? "production",
                bundleId: options["bundleId"] as? String,
                maxBatchSize: options["maxBatchSize"] as? Int ?? 100,
                flushInterval: options["flushInterval"] as? TimeInterval ?? 30,
                maxStoredEvents: options["maxStoredEvents"] as? Int ?? 10000,
                enableDebugLogging: options["enableDebugLogging"] as? Bool ?? false,
                trackAppLifecycleEvents: options["trackAppLifecycleEvents"] as? Bool ?? true
            )
        }

        MostlyGoodMetrics.configure(with: config)
    }

    @objc
    func track(_ name: String, properties: NSDictionary?) {
        let props = properties as? [String: Any]
        MostlyGoodMetrics.track(name, properties: props)
    }

    @objc
    func identify(_ userId: String) {
        MostlyGoodMetrics.identify(userId: userId)
    }

    @objc
    func resetIdentity() {
        MostlyGoodMetrics.shared?.resetIdentity()
    }

    @objc
    func flush() {
        MostlyGoodMetrics.flush()
    }

    @objc
    func startNewSession() {
        MostlyGoodMetrics.shared?.startNewSession()
    }

    @objc
    func clearPendingEvents() {
        MostlyGoodMetrics.shared?.clearPendingEvents()
    }

    @objc
    func getPendingEventCount(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let count = MostlyGoodMetrics.shared?.pendingEventCount ?? 0
        resolve(count)
    }
}
